### DISCUSSION TITLE
ModelsBuilder generates `IPublishedContent` or `IEnumerable<IPublishedContent>` for single allowed type

### DISCUSSION THREAD

**Discussion by bjarnef** [umbraco/Umbraco-CMS#15266] [Features and ideas]
I don't recall if this was different in past, but I noticed for MNTP configured as single picker and a specific type it generates `IPublishedContent` and if multipicker `IEnumerable<IPublishedContent>`.

If only a single content type is allowed, would it be possible to generate the specific type `T` or `IEnumerable<T>` instead? e.g. `NewsArticle` or `IEnumerable<NewsArticle>`?

**Comment by abjerner** [comment]
I think this will be a breaking change, and it might even be too late to fix for Umbraco 13 😢 

But I think it would make good sense to fix for the next major where we're still allowed to make breaking changes.

Also, the problem isn't really ModelsBuilder itself, but the property value type returned by the MNTP's property value converter.

For now, you can either use @callumbwhyte's [Super Value Converters](https://github.com/callumbwhyte/super-value-converters/) package or my own [Limbo Multinode Treepicker](https://github.com/abjerner/Limbo.Umbraco.MultiNodeTreePicker) (I may have had a look or two at Callum's code). Both packages also support compositions, so if you allow two content types sharing the same composition, the generated property value type will be `IMyFancyComposition` / `IEnumerable<IMyFancyComposition>` opposed to just `IPublishedContent` / `IEnumerable<IPublishedContent>`.

**Reply by bjarnef** [reply]
Yes, it would indeed be a breaking change.
Maybe I am wrong, I just recall it once did, or maybe only for value converter for Content Picker.

Furthermore with nullable types a single picker should probably also return `IPublishedContent?` or `T?`.
Related to this I think https://github.com/umbraco/Umbraco-CMS/discussions/13816 would be really useful (someday hopefully) 😎

I like the use of compositions - however sometimes I think it would be cool if the could override label and description to elaborate in the specific context https://github.com/umbraco/Umbraco-CMS/discussions/12967 e.g. a property on Page 1 using type A may not use the property exact same way on Page 2 using type B.
Maybe even override validation, e.g. mandatory on Page 1, but not on Page 2.

**Reply by abjerner** [reply]
I would love to see better use of nullability in ModelsBuilder. To be fair, I haven't looked much at the build in ModelsBuilder in U11 and U12 since we have our own ModelsBuilder package, so I'm not sure about how much has been added since U10. But for the build-in U10 ModelsBuilder, there isn't any nullabilty. Instead properties that where the type isn't a value type will get an `[MayBeNull]` attribute - even if the property type is a collection that ideally wouldn't be null.

Ideally a property value converter should be able to return the type of value returned by the property, but also whether said the returned value is nullable. Currently we're only able to indicate the the property value type, but not whether that type is nullable. I've tried looking for ways to do this, but I don't think it's possible without making changes to `IPropertyValueConverter` interface: https://github.com/umbraco/Umbraco-CMS/blob/contrib/src/Umbraco.Core/PropertyEditors/IPropertyValueConverter.cs#L42

It might however also we worth noticing that the generated models represent a point in time, so properties that would be known to never be null, would still return null if the property is removed from the content type, but the models aren't regenerated. Or you make other changes to the content types and data types, but don't regenerate the models. One could argue that a property in a generated model should then always be nullable. One could also argue that it be okay that these properties are not nullable - as you should always regenerate the models when making changes to content types and data types. If HQ were to add better support for nullables, they might need to decide on the best approach for this.

**Comment by skttl** [comment]
I've been wanting this forever - sorry for not speaking up :)

The sooner we can get this, the better.

**Reply by bjarnef** [reply]
It would indeed make the code cleaner in many use-cases.
Often we prefer MNTP over Content Picker since it allows a specific type and limit the start node.
In fact I am not sure we need Content Picker in future (it is based on same HTML. CSS and JS and MNTP), but more some kind of entity picker. I think @mattbrailsford used added a similar concept in [Konstrukt package ](https://docs.getkonstrukt.net/property-editors/entity-picker) now [Umbraco UI Builder](https://docs.umbraco.com/umbraco-ui-builder/property-editors/entity-picker).
